### PR TITLE
JavaScript dispatch pipeline

### PR DESCRIPTION
**Pull request by pepone**
This PR implements the new dispatch pipeline for JavaScript

### CODE REVIEW

File: js/src/Ice/CurrentExtensions.js [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+            replyStatus = ReplyStatus.ObjectNotExist;
+        } else if (exception instanceof FacetNotExistException) {
+            replyStatus = ReplyStatus.FacetNotExist;
+        } else if (exception instanceof OperationNotExistException) {
+            replyStatus = ReplyStatus.OperationNotExist;
+        } else {
+            throw new MarshalException("Unexpected exception type");
+        }
+
+        let id = exception.id;
```
Comment (bernardnormier) [inline js/src/Ice/CurrentExtensions.js:106]: This fixes #2345.

File: js/src/Ice/CurrentExtensions.js [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+        let facet = exception.facet;
+        if (id.name.length == 0) {
+            id = current.id;
+            facet = current.facet;
+        }
+        const operation = exception.operation.length == 0 ? current.operation : exception.operation;
+
+        exceptionDetails = ""; // TODO
```
Comment (bernardnormier) [inline js/src/Ice/CurrentExtensions.js:114]: Call RequestFailedException.createMessage.

File: js/src/Ice/Current.js [hunk 3]
```diff
@@ -2,7 +2,6 @@
 // Copyright (c) ZeroC, Inc. All rights reserved.
 //
 
-import { defineStruct } from "./Struct.js";
 import { Ice as Ice_OperationMode } from "./OperationMode.js";
```
Comment (bernardnormier) [inline js/src/Ice/Current.js:5]: The Current constructor below should not have defaulted parameters.

File: js/src/Ice/Current.d.ts [hunk 4]
```diff
@@ -34,6 +34,13 @@ declare module "ice" {
             encoding: EncodingVersion;
             static write(outs: OutputStream, value: Current): void;
             static read(ins: InputStream): Current;
+
```
Comment (bernardnormier) [inline js/src/Ice/Current.d.ts:37]: many of the ctor parameter above should not be nullable.

**Review by bernardnormier — APPROVED** [review]
Looks good.

**Review by externl — APPROVED** [review]

File: js/src/Ice/LocalExceptions.js [hunk 5]
```diff
@@ -18,7 +20,7 @@ export class RequestFailedException extends LocalException {
         return "::Ice::RequestFailedException";
     }
 
-    constructor(typeName, id, facet, operation) {
+    constructor(typeName, id = new Identity(), facet = "", operation = "") {
```
[the remark this answers is not in the document]
Comment (pepone) [reply js/src/Ice/LocalExceptions.js:23]: Yes, the reference to the imported class is the const above.

File: js/src/Ice/OutgoingResponse.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+    get exceptionId() {
+        return this._exceptionId;
+    }
+
+    get exceptionDetails() {
+        return this._exceptionDetails;
+    }
+
+    get size() {
```
[the remark this answers is not in the document]
Comment (pepone) [reply js/src/Ice/OutgoingResponse.js:34]: Yes I ported the OutgoingResponse for C#, didn't notice we don't actually use it in JavaScript.

File: js/src/Ice/LocalExceptions.js [hunk 7]
```diff
@@ -5,6 +5,8 @@
 import { LocalException } from "./LocalException.js";
 import { identityToString } from "./IdentityToString.js";
 import { ObjectPrx } from "./ObjectPrx.js";
+import { Ice as Ice_Identity } from "./Identity.js";
```
[the remark this answers is not in the document]
Comment (pepone) [reply js/src/Ice/LocalExceptions.js:8]: The generated code only exports an object for each top-level module. We don't actually use the generated code for Identity.js to work-around circular dependencies, but generated code that includes Identity.ice expect this.

**Pull request merged**